Analysis/LeptonID/src/PIDutil.cc

```diff
@@ -0,0 +1,374 @@
+        std::vector<double> thphcov(dir_error.begin(), dir_error.end());
+
+        return WeightedPoints3D(cogv, cov_fr_clu, thphcov, npoints, wgt_sum, wgt2_sum, wgt4_sum);
+    }
+}
+
+MCParticle *getMCParent(const MCParticle *mcp)
+{
+    MCParticle *parent = nullptr;
+
+    auto parents = mcp->getParents();
+    if (parents.size() > 0) {
+        if (parents[0]->getPDG() != mcp->getPDG())
+        {
+            parent = parents[0];
+        } else {
+            // recursively walk up the chain to the real parent
+            parent = getMCParent(parents[0]);
+        }
+    }
+
+    return parent;
+}
+
+std::vector<float> getPosMomFromTrackState(const TrackState *ts, double B) {
+    double trackPt = B * 3.0e-4 / std::fabs(ts->getOmega());
+    float trackPx = trackPt * std::cos(ts->getPhi());
+    float trackPy = trackPt * std::sin(ts->getPhi());
+    float trackPz = trackPt * ts->getTanLambda();
+
+    float trackXs = ts->getReferencePoint()[0] - ts->getD0() * std::sin(ts->getPhi());
+    float trackYs = ts->getReferencePoint()[1] + ts->getD0() * std::cos(ts->getPhi());
+    float trackZs = ts->getReferencePoint()[2] + ts->getZ0();
+    return std::vector<float>{trackXs, trackYs, trackZs, trackPx, trackPy, trackPz};
+}
+
+double distFun(const double *val, const double *par)
+{
+    double phi = val[0];
+    double v = val[1];
+
+    double x_ref = par[0];
+    double y_ref = par[1];
+    double z_ref = par[2];
+
+    double inv_Omega_minus_d_zero = par[3];
+
+    double sin_phi_zero = par[4];
+    double cos_phi_zero = par[5];
+
+    double z_zero = par[6];
+    double tan_lambda_over_Omega = par[7];
+    double phi_zero = par[8];
+    double inv_Omega = par[9];
+
+    double x_third = par[10];
+    double y_third = par[11];
+    double z_third = par[12];
+
+    double Px_third = par[13];
+    double Py_third = par[14];
+    double Pz_third = par[15];
+
+    double x_track = x_ref + inv_Omega_minus_d_zero * sin_phi_zero - inv_Omega * sin(phi);
+    double y_track = y_ref - inv_Omega_minus_d_zero * cos_phi_zero + inv_Omega * cos(phi);
+    double z_track = z_ref + z_zero - tan_lambda_over_Omega * (phi - phi_zero);
+
+    double x_line = x_third + Px_third * v;
+    double y_line = y_third + Py_third * v;
+    double z_line = z_third + Pz_third * v;
+
+    double dist_x_sq = (x_track - x_line) * (x_track - x_line);
+    double dist_y_sq = (y_track - y_line) * (y_track - y_line);
+    double dist_z_sq = (z_track - z_line) * (z_track - z_line);
+
+    double dist = sqrt(dist_x_sq + dist_y_sq + dist_z_sq);
+
+    return dist;
+}
+
+double intersectTrackLine(Track *track, TVector3 momentumOfLine,
+                          std::vector<double> pointOnLine,
+                          Vertex *primaryVertex, TVector3 &PCAatTrack,
+                          TVector3 &PCAatLine)
+{
+    double phi_min, phi_max, v_min, v_max;
+    phi_min = track->getPhi() - 3.14159265359 / 40.0;
+    phi_max = track->getPhi() + 3.14159265359 / 40.0;
```

Comment:
Can we give the magic number `40` a name here?

Reply:
Maybe. This whole function is more or less copy paste [from Yasser](https://github.com/yradkhorrami/SLDecayCorrection/blob/58f59925bc893cfe1abb17fa9a6bbac48b00af76/src/flightDirection.cc#L282-L436). But I think it is also not directly used in this PR.